apis/python/src/tiledb/vector_search/object_api/embeddings_ingestion.py

```diff
@@ -0,0 +1,313 @@
+from functools import partial
+from typing import Any, Dict, List, Mapping, Optional
+
+from tiledb.cloud.dag import Mode
+from tiledb.vector_search.object_api import ObjectIndex
+from tiledb.vector_search.object_readers import ObjectPartition
+
+
+def ingest_embeddings(
+    object_index: ObjectIndex,
+    embeddings_uri: str,
+    metadata_array_uri: str = None,
+    index_timestamp: int = None,
+    workers: int = -1,
+    worker_resources: Dict = None,
+    worker_image: str = None,
+    max_tasks_per_stage: int = -1,
+    verbose: bool = False,
+    trace_id: Optional[str] = None,
+    mode: Mode = Mode.LOCAL,
+    config: Optional[Mapping[str, Any]] = None,
+    namespace: Optional[str] = None,
+    **kwargs,
+):
+    import logging
+    import math
+    import time
+
+    import tiledb
+    from tiledb.cloud import dag
+    from tiledb.cloud.rest_api import models
+    from tiledb.cloud.utilities import get_logger
+    from tiledb.cloud.utilities import set_aws_context
+
+    MAX_TASKS_PER_STAGE = 100
+    DEFAULT_IMG_NAME = "3.9-vectorsearch"
+    DEFAULT_WORKER_RESOURCES = {"cpu": "1", "memory": "4Gi"}
+
+    def setup(
+        config: Optional[Mapping[str, Any]] = None,
+        verbose: bool = False,
+    ) -> logging.Logger:
+        """
+        Set the default TileDB context, OS environment variables for AWS,
+        and return a logger instance.
+
+        :param config: config dictionary, defaults to None
+        :param verbose: verbose logging, defaults to False
+        :return: logger instance
+        """
+
+        try:
+            tiledb.default_ctx(config)
+        except tiledb.TileDBError:
+            # Ignore error if the default context was already set
+            pass
+
+        set_aws_context(config)
+
+        level = logging.DEBUG if verbose else logging.NOTSET
+        logger = get_logger(level)
+
+        logger.debug(
+            "tiledb.cloud=%s, tiledb=%s, libtiledb=%s",
+            tiledb.cloud.version.version,
+            tiledb.version(),
+            tiledb.libtiledb.version(),
+        )
+
+        return logger
+
+    # --------------------------------------------------------------------
+    # UDFs
+    # --------------------------------------------------------------------
+    def compute_embeddings_udf(
+        object_reader_source_code: str,
+        object_reader_class_name: str,
+        object_reader_kwargs: Dict,
+        object_embedding_source_code: str,
+        object_embedding_class_name: str,
+        object_embedding_kwargs: Dict,
+        partition_dicts: List[Dict],
+        embeddings_uri: str,
+        metadata_array_uri: str = None,
+        index_timestamp: int = None,
+        verbose: bool = False,
+        trace_id: Optional[str] = None,
+        config: Optional[Mapping[str, Any]] = None,
+    ):
+        import numpy as np
+
+        import tiledb
+
+        def instantiate_object(code, class_name, **kwargs):
```

Comment:
This is a way to pass a user defined class(reader/embedding) to a UDF.

Pickle only packages the object fields and not its code. So if we packaged the objects the code would be missing on the UDF side and we would fail to initialize it. Some library readers/embeddings will be available in the UDF image but without this we would be unable to execute user defined readers for custom formats.

Solution:
- Client side: 
  - Read the source code of the custom class as string
  - Get the class name string
  - Package the init kwargs
  - Send (source code str, class name str, init kwargs)
- Server side:
  - Write the source code to a temp file
  - Load and import the temp file module
  - From the temp file module instantiate an object with class name and init kwargs